tests/federation_room_join_partial_state_test.go

```diff
@@ -177,7 +259,78 @@ func TestPartialStateJoin(t *testing.T) {
 		)
 		if err := matcher([]byte(roomRes.Raw)); err != nil {
 			t.Errorf("Did not find expected state events in /sync response: %s", err)
+		}
+	}
 
+	t.Run("EagerInitialSyncDuringPartialStateJoin", func(t *testing.T) {
+		eagerSyncDuringPartialStateJoinTest(t, "initial", false)
+	})
+
+	t.Run("EagerIncrementalSyncDuringPartialStateJoin", func(t *testing.T) {
+		eagerSyncDuringPartialStateJoinTest(t, "incremental", true)
+	})
+
+	// The tests above use long-polling syncs with a (complement-default) timeout of
+	// 1s. They don't test that the long-poll returns early when the partial state
+	// resync completes. This test does exactly that.
+	t.Run("EagerLongPollingSyncWokenWhenResyncCompletes", func(t *testing.T) {
+		alice := deployment.RegisterUser(t, "hs1", "t1alice_long_polling", "secret", false)
+
+		t.Log("Alice partial-joins a remote room.")
+		server := createTestServer(t, deployment)
+		cancel := server.Listen()
+		defer cancel()
+		serverRoom := createTestRoom(t, server, alice.GetDefaultRoomVersion(t))
+		psjResult := beginPartialStateJoin(t, server, serverRoom, alice)
+		defer psjResult.Destroy(t)
+
+		t.Log("Alice eager-syncs. The response should not contain the remote room.")
+		response, nextBatch := alice.MustSync(t, client.SyncReq{})
+
+		syncJoinedRoomPath := "rooms.join." + client.GjsonEscape(serverRoom.RoomID)
+		if response.Get(syncJoinedRoomPath).Exists() {
+			t.Fatal("Sync shouldn't include the joined room until resync is over")
+		}
+
+		// Begin a long polling sync that shouldn't return yet since no change happened
+		responseChan := make(chan gjson.Result, 1)
+		syncStarted := make(chan struct{})
+		go func() {
+			defer close(responseChan)
+			defer close(syncStarted)
+
+			syncStarted <- struct{}{}
+			response, _ := alice.MustSync(t, client.SyncReq{
+				TimeoutMillis: "10000",
+				Since:         nextBatch,
+			})
+			responseChan <- response
+		}()
+
+		// Try to wait for the sync to actually start, then un-partial-state the room
+		select {
+		case <-syncStarted:
+			// wait for the state_ids request to arrive
+			psjResult.AwaitStateIdsRequest(t)
+			// release the federation /state response
+			psjResult.FinishStateRequest()
+		case <-time.After(time.Second * 5):
+			// even though this should mostly be impossible, make sure we have a timeout
+			t.Fatalf("goroutine didn't start")
+		}
```

Comment:
This isn't a strong guarantee that the homeserver has received the /sync and started waiting for something to happen yet. It's possible for us to call `FinishStateRequest` before the /sync request is sent out.

I'd suggest
1. sleeping for a few seconds to confirm the homeserver is waiting (and maybe check `responseChan` for sanity)
2. then allowing the partial state join to complete
3. then expecting the sync response